File: src/api/shell/workspaces.py
"""Workspace CRUD routes."""

from __future__ import annotations

from typing import Optional

from fastapi import APIRouter, Depends, Query, status

from src.api.deps import get_workspace_service
from src.api.schemas.settings import WorkspaceSettingsOut
from src.api.schemas.workspace import (
    WorkspaceCreate,
    WorkspaceDetailOut,
    WorkspaceListOut,
    WorkspaceOut,
    WorkspaceUpdate,
)
from src.app.constants import DEFAULT_PAGE_LIMIT, MAX_PAGE_LIMIT
from src.app.entities.settings import WorkspaceSettingsEntity
from src.app.entities.workspace import WorkspaceEntity
from src.app.services.workspace_service import WorkspaceService

router = APIRouter(tags=["Workspaces"])


def _ws_out(ws: WorkspaceEntity) -> WorkspaceOut:
    return WorkspaceOut(
        id=ws.id,
        name=ws.name,
        path=ws.path,
        description=ws.description,
        status=ws.status,
        updated_at=ws.updated_at,
        deleted_at=ws.deleted_at,
    )


def _settings_out(s: WorkspaceSettingsEntity) -> WorkspaceSettingsOut:
    return WorkspaceSettingsOut(
        default_agent=s.default_agent,
        enabled_chatbots=s.enabled_chatbots,
        audience_default=s.audience_default,
        auto_index=s.auto_index,
        mcp_enabled=s.mcp_enabled,
    )


@router.get("/workspaces", response_model=WorkspaceListOut)
def list_workspaces(
    q: Optional[str] = None,
    status_filter: Optional[str] = Query(default=None, alias="status"),
    limit: int = Query(default=DEFAULT_PAGE_LIMIT, ge=1, le=MAX_PAGE_LIMIT),
    offset: int = Query(default=0, ge=0),
    svc: WorkspaceService = Depends(get_workspace_service),
) -> WorkspaceListOut:
    items, total = svc.list_active(q=q, status=status_filter, limit=limit, offset=offset)
    return WorkspaceListOut(
        items=[_ws_out(w) for w in items],
        total=total,
        limit=limit,
        offset=offset,
    )


@router.get("/workspaces/deleted", response_model=WorkspaceListOut)
def list_deleted_workspaces(
    q: Optional[str] = None,
    limit: int = Query(default=DEFAULT_PAGE_LIMIT, ge=1, le=MAX_PAGE_LIMIT),
    offset: int = Query(default=0, ge=0),
    svc: WorkspaceService = Depends(get_workspace_service),
) -> WorkspaceListOut:
    items, total = svc.list_deleted(q=q, limit=limit, offset=offset)
    return WorkspaceListOut(
        items=[_ws_out(w) for w in items],
        total=total,
        limit=limit,
        offset=offset,
    )


@router.get("/workspaces/{workspace_id}", response_model=WorkspaceDetailOut)
def get_workspace(
    workspace_id: str,
    svc: WorkspaceService = Depends(get_workspace_service),
) -> WorkspaceDetailOut:
    ws, settings = svc.get_detail(workspace_id)
    base = _ws_out(ws)
    return WorkspaceDetailOut(
        **base.model_dump(),
        created_at=ws.created_at,
        settings=_settings_out(settings),
    )


@router.post("/workspaces", response_model=WorkspaceOut, status_code=status.HTTP_201_CREATED)
def create_workspace(
    body: WorkspaceCreate,
    svc: WorkspaceService = Depends(get_workspace_service),
) -> WorkspaceOut:
    ws = svc.create(name=body.name, path=body.path, description=body.description)
    return _ws_out(ws)


@router.patch("/workspaces/{workspace_id}", response_model=WorkspaceOut)
def update_workspace(
    workspace_id: str,
    body: WorkspaceUpdate,
    svc: WorkspaceService = Depends(get_workspace_service),
) -> WorkspaceOut:
    ws = svc.update(
        workspace_id,
        name=body.name,
        path=body.path,
        description=body.description,
    )
    return _ws_out(ws)


@router.delete("/workspaces/{workspace_id}", response_model=WorkspaceOut)
def delete_workspace(
    workspace_id: str,
    svc: WorkspaceService = Depends(get_workspace_service),
) -> WorkspaceOut:
    return _ws_out(svc.soft_delete(workspace_id))


@router.post("/workspaces/{workspace_id}/restore", response_model=WorkspaceOut)
def restore_workspace(
    workspace_id: str,
    svc: WorkspaceService = Depends(get_workspace_service),
) -> WorkspaceOut:
    return _ws_out(svc.restore(workspace_id))

File: frontend/src/api/workspaces.ts
/** App shell workspace API — snake_case types matching backend. */

import { apiFetch } from "./client";

export type WorkspaceStatus = "ready" | "indexing" | "error" | "idle";
export type AudienceDefault = "end_user" | "technical";
export type UpdateJobStatus = "running" | "success" | "failed";
export type LogLevel = "info" | "warn" | "error";
export type ChatRole = "user" | "assistant";
export type ChatbotRole = "end_user" | "technical";

export interface WorkspaceOut {
  id: string;
  name: string;
  path: string;
  description: string;
  status: WorkspaceStatus;
  updated_at: string;
  deleted_at: string | null;
}

export interface WorkspaceSettings {
  default_agent: string;
  enabled_chatbots: string[];
  audience_default: AudienceDefault;
  auto_index: boolean;
  mcp_enabled: boolean;
}

export interface WorkspaceDetailOut extends WorkspaceOut {
  created_at: string;
  settings: WorkspaceSettings;
}

export interface WorkspaceInput {
  name: string;
  path: string;
  description: string;
}

export interface WorkspaceListOut {
  items: WorkspaceOut[];
  total: number;
  limit: number;
  offset: number;
}

export interface UpdateJob {
  id: string;
  started_at: string;
  finished_at: string | null;
  status: UpdateJobStatus;
  summary: string;
  detail: string;
}

export interface UpdateJobAccepted {
  job_id: string;
  status: UpdateJobStatus;
  started_at: string;
}

export interface UpdateJobListOut {
  items: UpdateJob[];
  total: number;
  limit: number;
  offset: number;
}

export interface ActivityLog {
  id: string;
  at: string;
  level: LogLevel;
  source: string;
  message: string;
}

export interface ActivityLogListOut {
  items: ActivityLog[];
  total: number;
  limit: number;
  offset: number;
}

export interface Chatbot {
  id: string;
  name: string;
  role: ChatbotRole | string;
}

export interface ChatMessage {
  id: string;
  role: ChatRole;
  text: string;
  at: string;
}

export interface ChatMessageListOut {
  items: ChatMessage[];
  total: number;
  limit: number;
  offset: number;
}

export interface SendMessageOut {
  user_message: ChatMessage;
  assistant_message: ChatMessage;
}

function qs(params: Record<string, string | number | undefined | null>): string {
  const sp = new URLSearchParams();
  for (const [k, v] of Object.entries(params)) {
    if (v === undefined || v === null || v === "") continue;
    sp.set(k, String(v));
  }
  const s = sp.toString();
  return s ? `?${s}` : "";
}

export function listWorkspaces(opts?: {
  q?: string;
  status?: string;
  limit?: number;
  offset?: number;
}): Promise<WorkspaceListOut> {
  return apiFetch(`/api/v1/workspaces${qs(opts || {})}`);
}

export function listDeletedWorkspaces(opts?: {
  q?: string;
  limit?: number;
  offset?: number;
}): Promise<WorkspaceListOut> {
  return apiFetch(`/api/v1/workspaces/deleted${qs(opts || {})}`);
}

export function getWorkspace(id: string): Promise<WorkspaceDetailOut> {
  return apiFetch(`/api/v1/workspaces/${encodeURIComponent(id)}`);
}

export function createWorkspace(input: WorkspaceInput): Promise<WorkspaceOut> {
  return apiFetch("/api/v1/workspaces", {
    method: "POST",
    body: JSON.stringify(input),
  });
}

export function updateWorkspace(
  id: string,
  input: Partial<WorkspaceInput>,
): Promise<WorkspaceOut> {
  return apiFetch(`/api/v1/workspaces/${encodeURIComponent(id)}`, {
    method: "PATCH",
    body: JSON.stringify(input),
  });
}

export function deleteWorkspace(id: string): Promise<WorkspaceOut> {
  return apiFetch(`/api/v1/workspaces/${encodeURIComponent(id)}`, {
    method: "DELETE",
  });
}

export function restoreWorkspace(id: string): Promise<WorkspaceOut> {
  return apiFetch(`/api/v1/workspaces/${encodeURIComponent(id)}/restore`, {
    method: "POST",
  });
}

export function getSettings(workspaceId: string): Promise<WorkspaceSettings> {
  return apiFetch(`/api/v1/workspaces/${encodeURIComponent(workspaceId)}/settings`);
}

export function putSettings(
  workspaceId: string,
  settings: WorkspaceSettings,
): Promise<WorkspaceSettings> {
  return apiFetch(`/api/v1/workspaces/${encodeURIComponent(workspaceId)}/settings`, {
    method: "PUT",
    body: JSON.stringify(settings),
  });
}

export function startUpdate(
  workspaceId: string,
  body?: { rebuild?: boolean; scope?: "full" | "incremental" },
): Promise<UpdateJobAccepted> {
  return apiFetch(`/api/v1/workspaces/${encodeURIComponent(workspaceId)}/update`, {
    method: "POST",
    body: JSON.stringify(body || { rebuild: true, scope: "full" }),
  });
}

export function listUpdates(
  workspaceId: string,
  opts?: { limit?: number; offset?: number },
): Promise<UpdateJobListOut> {
  return apiFetch(
    `/api/v1/workspaces/${encodeURIComponent(workspaceId)}/updates${qs(opts || {})}`,
  );
}

export function getUpdate(workspaceId: string, jobId: string): Promise<UpdateJob> {
  return apiFetch(
    `/api/v1/workspaces/${encodeURIComponent(workspaceId)}/updates/${encodeURIComponent(jobId)}`,
  );
}

export function listLogs(
  workspaceId: string,
  opts?: { level?: string; q?: string; limit?: number; offset?: number },
): Promise<ActivityLogListOut> {
  return apiFetch(
    `/api/v1/workspaces/${encodeURIComponent(workspaceId)}/logs${qs(opts || {})}`,
  );
}

export function listChatbots(workspaceId: string): Promise<{ items: Chatbot[] }> {
  return apiFetch(`/api/v1/workspaces/${encodeURIComponent(workspaceId)}/chatbots`);
}

export function listMessages(
  workspaceId: string,
  botId: string,
  opts?: { limit?: number; offset?: number },
): Promise<ChatMessageListOut> {
  return apiFetch(
    `/api/v1/workspaces/${encodeURIComponent(workspaceId)}/chatbots/${encodeURIComponent(botId)}/messages${qs(opts || {})}`,
  );
}

export function sendMessage(
  workspaceId: string,
  botId: string,
  text: string,
): Promise<SendMessageOut> {
  return apiFetch(
    `/api/v1/workspaces/${encodeURIComponent(workspaceId)}/chatbots/${encodeURIComponent(botId)}/messages`,
    {
      method: "POST",
      body: JSON.stringify({ text }),
    },
  );
}

/** Poll an update job until it leaves `running`, or attempts are exhausted. */
export async function pollUpdateJob(
  workspaceId: string,
  jobId: string,
  opts?: { intervalMs?: number; maxAttempts?: number },
): Promise<UpdateJob> {
  const intervalMs = opts?.intervalMs ?? 1500;
  const maxAttempts = opts?.maxAttempts ?? 60;
  let last = await getUpdate(workspaceId, jobId);
  for (let i = 0; i < maxAttempts && last.status === "running"; i++) {
    await new Promise((r) => setTimeout(r, intervalMs));
    last = await getUpdate(workspaceId, jobId);
  }
  return last;
}
